mono/profiler/proflog.c

```diff
@@ -3894,11 +3995,16 @@ log_shutdown (MonoProfiler *prof)
 	}
 #endif
 
-	if (TLS_GET (LogBuffer, tlsbuffer))
-		send_buffer (prof, TLS_GET (GPtrArray, tlsmethodlist), TLS_GET (LogBuffer, tlsbuffer));
-
-	TLS_SET (tlsbuffer, NULL);
-	TLS_SET (tlsmethodlist, NULL);
+	/*
+	 * Ensure that we empty the LLS completely, even if some nodes are
+	 * not immediately removed upon calling mono_lls_remove (), by
+	 * iterating until the head is NULL.
+	 */
+	while (profiler_thread_list.head) {
```

Comment:
You might be right. I'd forgotten about threads like the finalizer thread and TP threads. Will have a closer look.